shell/platform/darwin/macos/framework/Source/FlutterTextInputPlugin.mm

```diff
@@ -620,6 +634,28 @@ - (void)doCommandBySelector:(SEL)selector {
     void (*func)(id, SEL, id) = reinterpret_cast<void (*)(id, SEL, id)>(imp);
     func(self, selector, nil);
   }
+
+  // Group multiple selectors received within a single run loop turn so that
+  // framework can process them in single microtask.
+  NSString* name = NSStringFromSelector(selector);
+  if (_pendingSelectors == nil) {
+    _pendingSelectors = [NSMutableArray array];
+  }
+  [_pendingSelectors addObject:name];
+  __weak NSMutableArray* selectors = _pendingSelectors;
+  __weak FlutterMethodChannel* channel = _channel;
+  __weak NSNumber* clientID = self.clientID;
+
+  CFStringRef runLoopMode = self.customRunLoopMode != nil
+                                ? (__bridge CFStringRef)self.customRunLoopMode
+                                : kCFRunLoopCommonModes;
+
+  CFRunLoopPerformBlock(CFRunLoopGetMain(), runLoopMode, ^{
```

Comment:
> Will this cause some events to reorder, for example, if a key is bound to 3 selectors [moveBackward, insert: a, moveForward], since insert:a will be handled by the input plugin, this essentially becomes [insert: a, moveBackward, moveForward], from the framework's perspective?

This is a good point. I'll test this. 

> Maybe use dispatch_async instead of CFRunLoopPerformBlock? Then you don't have to set the mode?

We can't pump the dispatch queue manually in unit tests. We can however run the run loop in custom mode so that only these blocks are executed.

Reply:
> Will this cause some events to reorder, for example, if a key is bound to 3 selectors [moveBackward, insert: a, moveForward], since insert:a will be handled by the input plugin, this essentially becomes [insert: a, moveBackward, moveForward], from the framework's perspective?

So it will indeed become that. However even if we flush buffered selectors and it will become [moveBackward, insert:a, moveForward] it still won't get expected result, because when `TextInputPlugin` processes `insert:a` the selection hasn't been updated by the framework yet so that first [moveBackward:] will only update selection after text is already inserted.

So I don't think we can handle compound actions that involve inserting text correctly at this point. It may required some rather significant changes to get this working.

Reply:
> We can't pump the dispatch queue manually in unit tests. We can however run the run loop in custom mode so that only these blocks are executed.

I thought we could add a dispatch barrier and use the async expectation API to verify the side effects?

Reply:
Something still needs to pump the main dispatch queue. I don't know any other way to do this except running the [CFRunLoop in default mode](https://github.com/opensource-apple/CF/blob/3cc41a76b1491f50813e28a4ec09954ffa359e6f/CFRunLoop.c#L1590). However running the `CFRunLoop` in default mode in tests crashes because of various scheduled `NSView` related tasks from other tests.

Dispatch barrier is unlikely to help here. Doing anything blocking on main thread (`dispatch_sync`, `dispatch_barrier_sync`, `dispatch_barrier_async_and_wait`) will crash with "LIBDISPATCH: dispatch_sync called on queue already owned by current thread". And async barrier will do not anything useful (it still needs the queue to be pumped).

Reply:
This test passes (didn't need a barrier since the main queue is not concurrent):
```swift
    func testExample() throws {
      var x = 0
      let expectation = XCTestExpectation()
      DispatchQueue.main.async {
        print("execute")
        x += 1
      }
      DispatchQueue.main.async {
        XCTAssertEqual(x, 1)
        expectation.fulfill()
      }
      print("wait")
      wait(for: [expectation], timeout: 10)
    }
```

Reply:
`xctwaiter.wait` calls `_dispatch_main_queue_drain()`, so the main queue is pumped in a sense.

Reply:
`_dispatch_main_queue_drain()` doesn't seem to be exported from libdispatch. Looking at XCTest it seems that `waitFor` actually [runs the run loop](https://github.com/apple/swift-corelibs-xctest/blob/dd599d04b0d17e5be3de5fd31101e7edda5d2b42/Sources/XCTest/Public/Asynchronous/XCTWaiter.swift#L367), which would have same sideeffect as calling `CFRunLoopRunInMode`with default modes manually, as in it would crash `flutter_desktop_darwin_unittests`. Now why this crash happens is another issue (perhaps we could drain the run loop in each test while the appropriate cocoa objects still exist), but that should be tracked as another issue.